### PR TITLE
Add support for data block hash index

### PR DESCRIPTION
**Pull request by dvdplm**
A good description of the feature enabled by this is found [here](https://rocksdb.org/blog/2018/08/23/data-block-hash-index.html).

Note: the librocksdb used by `rust-rocksdb` does not expose the necessary functions in the C api; pending PR upstream: facebook/rocksdb#6101. EDIT: merged.

- [x] update upstream rocksdb (needs v6.7.x)
- [x] add test (requires facebook/rocksdb#6101)

~~Currently blocked on https://github.com/rust-rocksdb/rust-rocksdb/pull/379https://github.com/rust-rocksdb/rust-rocksdb/pull/398~~

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -161,6 +161,18 @@ pub enum BlockBasedIndexType {
     TwoLevelIndexSearch,
 }
 
+/// Used by BlockBasedOptions::set_data_block_index_type.
+pub enum DataBlockIndexType {
```
Comment (ordian) [inline src/lib.rs]:
```suggestion
#[repr(C)]
pub enum DataBlockIndexType {
```

File: src/lib.rs [hunk 2]
```diff
@@ -161,6 +161,18 @@ pub enum BlockBasedIndexType {
...
+    /// Use binary search when performing point lookup for keys in data blocks.
+    /// This is the default.
+    BinarySearch,
```
Comment (ordian) [inline src/lib.rs]:
```suggestion
    BinarySearch = 0,
```

File: src/lib.rs [hunk 3]
```diff
@@ -161,6 +161,18 @@ pub enum BlockBasedIndexType {
...
+
+    /// Appends a compact hash table to the end of the data block for efficient indexing. Backwards
+    /// compatible with databases created without this feature. Once turned on, existing data will
+    /// be gradually converted to the hash index format.
+    BinaryAndHash,
```
Comment (ordian) [inline src/lib.rs]:
```suggestion
    BinaryAndHash = 1,
```

File: src/lib.rs (hunk 1 above)
Comment (dvdplm) [reply src/lib.rs]: Would this let me avoid doing [this](https://github.com/rust-rocksdb/rust-rocksdb/pull/358/files#diff-915d9d8a57469f27ff59479e2b15dba1R208)?

Comment (ordian) [reply src/lib.rs]: no, but I think the default repr for enums is `isize` https://github.com/rust-lang/rust/issues/24290,
so even though in practice it doesn't cause any trouble, we should be the safe side with `[repr(C)]` or `#[repr(i32)]`

**Comment by ordian** [comment]
please merge with master

**Comment by dvdplm** [comment]
@aleksuss PTAL

**Pull request merged**